source/windows/bcrypt_rsa.c

```diff
@@ -0,0 +1,458 @@
+     * seems to have functionality we need.
+     */
+    AWS_ZERO_STRUCT(key_impl->base.priv);
+    AWS_ZERO_STRUCT(key_impl->base.pub);
+
+    key_impl->base.vtable = &s_rsa_key_pair_vtable;
+    key_impl->base.key_size_in_bits = key_size_in_bits;
+
+    return &key_impl->base;
+
+on_error:
+    s_rsa_destroy_key(&key_impl->base);
+    return NULL;
+}
+
+struct aws_rsa_key_pair *aws_rsa_key_pair_new_from_private_key_pkcs1_impl(
+    struct aws_allocator *allocator,
+    struct aws_byte_cursor key) {
+
+    aws_thread_call_once(&s_rsa_thread_once, s_load_alg_handle, NULL);
+    struct bcrypt_rsa_key_pair *key_pair_impl = aws_mem_calloc(allocator, 1, sizeof(struct bcrypt_rsa_key_pair));
+
+    aws_ref_count_init(&key_pair_impl->base.ref_count, &key_pair_impl->base, aws_rsa_key_pair_destroy);
+    key_pair_impl->base.impl = key_pair_impl;
+    key_pair_impl->base.allocator = allocator;
+    aws_byte_buf_init_copy_from_cursor(&key_pair_impl->base.priv, allocator, key);
+
+    struct aws_der_decoder *decoder = aws_der_decoder_new(allocator, key);
+
+    if (!decoder) {
+        goto on_error;
+    }
+
+    struct aws_rsa_private_key_pkcs1 private_key_data;
+    AWS_ZERO_STRUCT(private_key_data);
+    if (aws_der_decoder_load_private_rsa_pkcs1(decoder, &private_key_data)) {
+        goto on_error;
+    }
+
+    /* Hard to predict final blob size, so use pkcs1 key size as upper bound. */
+    size_t total_buffer_size = key.len + sizeof(BCRYPT_RSAKEY_BLOB);
+
+    if (aws_byte_buf_init(&key_pair_impl->key_buf, allocator, total_buffer_size)) {
+        goto on_error;
+    }
+
+    aws_byte_buf_secure_zero(&key_pair_impl->key_buf);
+
+    BCRYPT_RSAKEY_BLOB key_blob;
+    AWS_ZERO_STRUCT(key_blob);
+    key_blob.Magic = BCRYPT_RSAFULLPRIVATE_MAGIC;
+    key_blob.BitLength = (ULONG)private_key_data.modulus.len * 8;
+    key_blob.cbPublicExp = (ULONG)private_key_data.publicExponent.len;
+    key_blob.cbModulus = (ULONG)private_key_data.modulus.len;
+    key_blob.cbPrime1 = (ULONG)private_key_data.prime1.len;
+    key_blob.cbPrime2 = (ULONG)private_key_data.prime2.len;
+
+    struct aws_byte_cursor header = aws_byte_cursor_from_array(&key_blob, sizeof(key_blob));
+    aws_byte_buf_append(&key_pair_impl->key_buf, &header);
+
+    LPCWSTR blob_type = BCRYPT_RSAFULLPRIVATE_BLOB;
+    ULONG flags = 0;
+
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.publicExponent);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.modulus);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.prime1);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.prime2);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.exponent1);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.exponent2);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.coefficient);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.privateExponent);
```

Comment:
I notice up above you told the blob about the length of 4 cursors:
cbPublicExp, cbModulus, cbPrime1, cbPrime2

but down here you write out 8 cursors

Just wanted to double-check with you that this is all correct. I'm not familiar enough with all this stuff to know if those last 4 cursors are constant lengths, or based on BitLength, or what

Reply:
yes, some of those have the same size. https://learn.microsoft.com/en-us/windows/win32/api/bcrypt/ns-bcrypt-bcrypt_rsakey_blob.

this is full rsa blob so it should have following components
```
BCRYPT_RSAKEY_BLOB
PublicExponent[cbPublicExp] // Big-endian.
Modulus[cbModulus] // Big-endian.
Prime1[cbPrime1] // Big-endian.
Prime2[cbPrime2] // Big-endian.
Exponent1[cbPrime1] // Big-endian.
Exponent2[cbPrime2] // Big-endian.
Coefficient[cbPrime1] // Big-endian.
PrivateExponent[cbModulus] // Big-endian.
```